### PR TITLE
New Module: win_dhcp_server

### PR DESCRIPTION
**Pull request by sgruber94**
### SUMMARY
The win_dhcp_server module manages Windows Server DHCP Options

### ISSUE TYPE
New Module Pull Request

### COMPONENT NAME
win_dhcp_server 

### ADDITIONAL INFORMATION
Includes Integration Tests - See documentation file for full details on how the module works.

### CODE REVIEW

File: plugins/modules/win_dhcp_server.ps1 [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+# SPDX-License-Identifier: GPL-3.0-only
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#AnsibleRequires -CSharpUtil Ansible.Basic
+$spec = @{
+    options             = @{
+        type              = @{ type = "str"; choices = "scope", "exclusion"; required =$true }
+        version           = @{ type = "str"; choices = "IPv4", "IPv6"; default = "IPv4" }
+        scope             = @{ type = "str"; aliases = "prefix" }
+        name              = @{ type = "str"; required =$true  }
+        validlifetime     = @{ type = "str" ; default = "0.04:00:00" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:12]: Try and split up multiple works with `_`, .e.g. `valid_lifetime`.

File: plugins/modules/win_dhcp_server.ps1 [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        preferredlifetime = @{ type = "str" ; default = "0.02:00:00" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:13]: Split with `_`, `preferred_lifetime`

File: plugins/modules/win_dhcp_server.ps1 [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        startrange        = @{ type = "str" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:14]: `start_range`

File: plugins/modules/win_dhcp_server.ps1 [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        endrange          = @{ type = "str" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:15]: `end_range`

File: plugins/modules/win_dhcp_server.ps1 [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        subnetmask        = @{ type = "str" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:16]: `subnet_mask`

File: plugins/modules/win_dhcp_server.ps1 [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        state             = @{ type = "str"; choices = "absent", "present"; default = "present" }
+        computername      = @{ type = "str" }
+        force             = @{ type = "bool"; default = $false }
+        scopestate        = @{ type = "str"; choices = "Active", "InActive"; default = "Active" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:20]: `scope_state`

File: plugins/modules/win_dhcp_server.ps1 [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        version           = @{ type = "str"; choices = "IPv4", "IPv6"; default = "IPv4" }
+        scope             = @{ type = "str"; aliases = "prefix" }
+        name              = @{ type = "str"; required =$true  }
+        validlifetime     = @{ type = "str" ; default = "0.04:00:00" }
+        preferredlifetime = @{ type = "str" ; default = "0.02:00:00" }
+        startrange        = @{ type = "str" }
+        endrange          = @{ type = "str" }
+        subnetmask        = @{ type = "str" }
+        state             = @{ type = "str"; choices = "absent", "present"; default = "present" }
+        computername      = @{ type = "str" }
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:18]: This is probably better to name `dhcp_server` as `computername` is a big vague. We do the same thing for the AD modules where the `-ComputerName` equivalent for those modules is mapped to the `domain_server` module option.

File: plugins/modules/win_dhcp_server.ps1 [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        if ($current_scope_exists -eq $false) {
+            $module.Result.msg = "The scope doesn't exist."
+        } else {
+            # If the scope exists, we need to destroy it
+            if ($version -eq "IPv4") {
+                try {
+                    Remove-DhcpServerv4Scope -ScopeId $scope_id @extra_param
+                    $module.Result.changed = $true
+                } catch {
+                    $module.Result.changed = $false
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:96]: No need to set back to false and really no need for the catch block, just let the exceptions flow.

File: plugins/modules/win_dhcp_server.ps1 [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                    $module.FailJson("Unable to remove scope : $($_.Exception.Message)")
+                }
+            } else {
+                #IPv6
+                try {
+                    Remove-DhcpServerv6Scope -Prefix $scope_id @extra_param
+                    $module.Result.changed = $true
+                } catch {
+                    $module.Result.changed = $false
```
Comment (jborean93) [inline plugins/modules/win_dhcp_server.ps1:105]: Remove catch block

File: tests/integration/targets/win_dhcp_server/tasks/tests.yml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+---
```
Comment (jborean93) [inline tests/integration/targets/win_dhcp_server/tasks/tests.yml:1]: Great to see the tests but you aren't really doing anything with them. You should be checking the return values, i.e. it reported a change (or not for idempotency) and checking the actual DHCP server to see if it added the scopes/done the action you specified.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: tests/integration/targets/win_dhcp_server/tasks/tests.yml (hunk 10 above)
Comment (sgruber94) [reply tests/integration/targets/win_dhcp_server/tasks/tests.yml:1]: I will add some test behaviour
should I add also the  version_added: '1.4.0'  to the documentation?

**Comment by jborean93** [comment]
See https://github.com/ansible-collections/community.windows/pull/194#issuecomment-877442730

**Pull request closed without merging**